File: Solutions/206.reverse-linked-list.cs
/*
 * @lc app=leetcode id=206 lang=csharp
 *
 * [206] Reverse Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

using CommonLibrary;

namespace _206;

public class Solution
{
    public ListNode ReverseList(ListNode head)
    {
        if (head == null) {
            return null;
        }
        if (head.next == null) {
            return head;
        }

        ListNode next_head = this.ReverseList(head.next);
        ListNode temp_tail = head.next;
        head.next = null;
        temp_tail.next = head;

        return next_head;
    }
}
// @lc code=end

File: Solutions/206.reverse-linked-list.cpp
/*
 * @lc app=leetcode id=206 lang=cpp
 *
 * [206] Reverse Linked List
 */
#include "common_header.hpp"

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution
{
public:
    ListNode* reverseList(ListNode* head)
    {
        if (head == nullptr) {
            return nullptr;
        }
        if (head->next == nullptr) {
            return head;
        }

        auto next_head = reverseList(head->next);
        auto temp_tail = head->next;
        head->next = nullptr;
        temp_tail->next = head;

        return next_head;
    }
};
// @lc code=end
